---
title: "H3K27me3 analysis using ChIPseeker"
author: "Clara Benoit-Pilven"
date: "`r Sys.Date()`"
output: html_document
---

```{r load_library, include=FALSE}
suppressPackageStartupMessages( library(readxl) )
suppressPackageStartupMessages( library(dplyr) )
suppressPackageStartupMessages( library(ggplot2) )
suppressPackageStartupMessages( library(tidyr) )
suppressPackageStartupMessages( library(ChIPseeker) )
suppressPackageStartupMessages( library(TxDb.Hsapiens.UCSC.hg38.knownGene) )
suppressPackageStartupMessages( library(org.Hs.eg.db) )
```

```{r set_workDir, echo=FALSE}
setwd("/Users/benoicla/Desktop/Workspace/iPSC/draft/revision/review_analyses/H3K27me3_analysis/")
```

# Load data

## Load HipSci sex-DE genes

Load the 3 sets of sex-DE genes.

```{r load_hipsci_sexDE, echo=FALSE}
sexDE_MG1 <-  read_excel("Table_S7_toptable_ipsc_MG1.xlsx", sheet = 1)
sexDE_MG1_Xchr <- sexDE_MG1[which(sexDE_MG1$chr == "X"),]
sexDE_MG1_Xchr$ensemblID <- sapply(strsplit(as.character(sexDE_MG1_Xchr$gene_id), "\\."), function(x){x[1]})
sexDE_MG2 <-  read_excel("Table_S8_toptable_ipsc_MG2.xlsx", sheet = 1)
sexDE_MG2_Xchr <- sexDE_MG2[which(sexDE_MG2$chr == "X"),]
sexDE_MG2_Xchr$ensemblID <- sapply(strsplit(as.character(sexDE_MG2_Xchr$gene_id), "\\."), function(x){x[1]})
sexDE_MG3 <-  read_excel("Table_S9_toptable_ipsc_MG3.xlsx", sheet = 1)
sexDE_MG3_Xchr <- sexDE_MG3[which(sexDE_MG3$chr == "X"),]
sexDE_MG3_Xchr$ensemblID <- sapply(strsplit(as.character(sexDE_MG3_Xchr$gene_id), "\\."), function(x){x[1]})
```

Also load the different comparison betyween female groups.

```{r load_hipsci_femaleGrpDE, echo=FALSE}
DE_G1G3 <- read_excel("toptable_ipsc_G1G3.xlsx", sheet = 1)
DE_G1G3_Xchr <- DE_G1G3[which(DE_G1G3$chr == "X"),]
DE_G1G3_Xchr$ensemblID <- sapply(strsplit(as.character(DE_G1G3_Xchr$gene_id), "\\."), function(x){x[1]})
DE_G1G2 <- read_excel("toptable_ipsc_G1G2.xlsx", sheet = 1)
DE_G1G2_Xchr <- DE_G1G2[which(DE_G1G2$chr == "X"),]
DE_G1G2_Xchr$ensemblID <- sapply(strsplit(as.character(DE_G1G2_Xchr$gene_id), "\\."), function(x){x[1]})
DE_G3G2 <- read_excel("toptable_ipsc_G3G2.xlsx", sheet = 1)
DE_G3G2_Xchr <- DE_G3G2[which(DE_G3G2$chr == "X"),]
DE_G3G2_Xchr$ensemblID <- sapply(strsplit(as.character(DE_G3G2_Xchr$gene_id), "\\."), function(x){x[1]})
```

In this analysis, we are only interested in what is happenning on the X-chromosome.

## Load histone marks using ChIPseeker

Annotate peaks.

```{r annotate_peaks, echo=FALSE}
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
K27me3_normal_annot <- annotatePeak("Data_Vallot_2015/GSM1528885_H9_XISTpos_K27me3.peaks.tab.bed",
                                    tssRegion=c(-3000, 3000), TxDb=txdb, annoDb="org.Hs.eg.db")
K27me3_eroded_annot <- annotatePeak("Data_Vallot_2015/GSM1528886_H9_XISTneg_K27me3.peaks.tab.bed",
                                    tssRegion=c(-3000, 3000), TxDb=txdb, annoDb="org.Hs.eg.db")
K9me3_normal_annot <- annotatePeak("Data_Vallot_2015/GSM1528888_H9_XISTpos_K9me3.peaks.tab.bed",
                                    tssRegion=c(-3000, 3000), TxDb=txdb, annoDb="org.Hs.eg.db")
K9me3_eroded_annot <- annotatePeak("Data_Vallot_2015/GSM1528889_H9_XISTneg_K9me3.peaks.tab.bed",
                                    tssRegion=c(-3000, 3000), TxDb=txdb, annoDb="org.Hs.eg.db")
```

Keep only peaks in X chromosome.

```{r Xchr_peaks, echo=FALSE}
K27me3_normal_annot_Xchr <- as.data.frame(K27me3_normal_annot)
K27me3_normal_annot_Xchr <- K27me3_normal_annot_Xchr[which(K27me3_normal_annot_Xchr$seqnames == "chrX"), ]
dim(K27me3_normal_annot_Xchr)
K27me3_eroded_annot_Xchr <- as.data.frame(K27me3_eroded_annot)
K27me3_eroded_annot_Xchr <- K27me3_eroded_annot_Xchr[which(K27me3_eroded_annot_Xchr$seqnames == "chrX"), ]
dim(K27me3_eroded_annot_Xchr)
K9me3_normal_annot_Xchr <- as.data.frame(K9me3_normal_annot)
K9me3_normal_annot_Xchr <- K9me3_normal_annot_Xchr[which(K9me3_normal_annot_Xchr$seqnames == "chrX"), ]
dim(K9me3_normal_annot_Xchr)
K9me3_eroded_annot_Xchr <- as.data.frame(K9me3_eroded_annot)
K9me3_eroded_annot_Xchr <- K9me3_eroded_annot_Xchr[which(K9me3_eroded_annot_Xchr$seqnames == "chrX"), ]
dim(K9me3_eroded_annot_Xchr)
```


# Enrichment analysis of histone marks

## Format the annotated peaks

Several formatting steps are needed before using this data in the enrichment analysis.

1. Keep only peaks in Promoter, 5'UTR, 3'UTR and exon.
2. Keep only peaks annotated with genes analyzed in our analyses


```{r format_peaks, echo=FALSE}
format_peaks <- function(peaks, DEdata = sexDE_MG1_Xchr, annotToKeep = c("Promoter", "5' UTR", "3' UTR", "Exon", "Intron")){
  peaks$annotationSimplified <- gsub(" \\(.*", "", peaks$annotation)
  peaks_filtered <- peaks[which(peaks$annotationSimplified %in% annotToKeep), ]
  peaks_filtered <- peaks_filtered[which(peaks_filtered$ENSEMBL %in% unique(DEdata$ensemblID)), ]
  peaks_final <- unique(peaks_filtered$ENSEMBL)
  return(peaks_final)
}
K27me3_normal_annot_Xchr_filtered <- format_peaks(K27me3_normal_annot_Xchr)
length(K27me3_normal_annot_Xchr_filtered)
K27me3_eroded_annot_Xchr_filtered <- format_peaks(K27me3_eroded_annot_Xchr)
length(K27me3_eroded_annot_Xchr_filtered)
K9me3_normal_annot_Xchr_filtered <- format_peaks(K9me3_normal_annot_Xchr)
length(K9me3_normal_annot_Xchr_filtered)
K9me3_eroded_annot_Xchr_filtered <- format_peaks(K9me3_eroded_annot_Xchr)
length(K9me3_eroded_annot_Xchr_filtered)
```
## Functions

Create a function that take into input: 
- the list genes with the DE results, 
- the list of genes for which we want to know the enrichment,
- the p-value significant threshold [default = 0.05],
- the direction of effect (in the case we want to look at up and down-regulated genes separately) [default = both]
And compute the enrichment and depletion of the gene list in the DE results.

```{r function_enrichment_analysis, echo=FALSE}
enrichment_analysis <- function(DEresults, geneList, analysisName, pval_threshold = 0.05, direction = "both"){
  if (direction == "both"){
    DEresults_signif <- DEresults[which(DEresults$adj.P.Val < pval_threshold), ]
  } else if (direction == "up"){
    DEresults_signif <- DEresults[which(DEresults$adj.P.Val < pval_threshold & DEresults$logFC > 0), ]
  } else if (direction == "down"){
    DEresults_signif <- DEresults[which(DEresults$adj.P.Val < pval_threshold & DEresults$logFC < 0), ]
  }
  universe_nbr <- length(DEresults$ensemblID)
  DE_nbr <- length(DEresults_signif$ensemblID)
  geneList_nbr <- length(geneList)
  # overlap
  overlap_nbr <- length(intersect(DEresults_signif$ensemblID, geneList))
  # get enrichment stat and effect size
  enrichment_stat_res <- enrichment_stat(overlap_nbr, geneList_nbr, DE_nbr, universe_nbr)
  # format data
  results <- data.frame("analysis" = analysisName, "direction" = direction, 
                        "nbr_DE" = DE_nbr, "nbr_geneList" = geneList_nbr, 
                        "overlap" = overlap_nbr, "N" = universe_nbr, 
                        "pval_enrich" = enrichment_stat_res$pvalEnrich, "pval_deple" = enrichment_stat_res$pvalDeple,
                        "relative_enrich" = enrichment_stat_res$relatEnrich)
  return(results)
}

enrichment_stat <- function(overlap_nbr, geneList_nbr, DE_nbr, universe_nbr){
  # overall enrichment
  #print(">> Enrichment")
  pvalEnrich <- phyper(overlap_nbr-1, geneList_nbr, universe_nbr - geneList_nbr, DE_nbr, lower.tail = FALSE)
  # overall depletion
  #print(">> Depletion")
  pvalDeple <-phyper(overlap_nbr,  geneList_nbr, universe_nbr -  geneList_nbr, DE_nbr, lower.tail = TRUE)
  # effect size
  #print(">> Relative enrichment")
  relatEnrich <- (overlap_nbr/ geneList_nbr)/(DE_nbr/universe_nbr)
  
  return(list("pvalEnrich" = pvalEnrich, "pvalDeple" = pvalDeple, "relatEnrich" = relatEnrich))
}
```

Function that do the enrichment analysis for the geneList of interest and a choosen number of random genesets.

```{r function_enrichment_analysis_permut, echo=FALSE}
enrichment_analysis_permut <- function(DEresults, geneList, analysisName, nbr_permut = 1000, pval_threshold = 0.05, direction = "both"){
  # enrichment for geneList
  results_enrich_geneList <- enrichment_analysis(DEresults = DEresults, geneList = geneList, 
                                                 analysisName = analysisName, pval_threshold = pval_threshold,
                                                 direction = direction)
  # enrichment for 'nbr_permut" random genesets of the same size of the geneList
  length_geneList <- length(geneList)
  results_enrich_randomGeneset <- results_enrich_geneList[0,]
  for (i in 1:nbr_permut){
    # select 'length_geneList' random genes from the universe
    random_geneset <- sample(DEresults$ensemblID, length_geneList)
    # enrichment for this random geneset
    results_enrich_randomGeneset_tmp <- enrichment_analysis(DEresults = DEresults, geneList = random_geneset, 
                                                            analysisName = paste(analysisName, "random", sep = "_"), 
                                                            pval_threshold = pval_threshold, direction = direction)
    results_enrich_randomGeneset <- rbind(results_enrich_randomGeneset, results_enrich_randomGeneset_tmp)
  }
  results_enrich_geneList$pval_enrich_random <- paste(results_enrich_randomGeneset$pval_enrich, sep = ";", collapse = ";")
  results_enrich_geneList$pval_deple_random <- paste(results_enrich_randomGeneset$pval_deple, sep = ";", collapse = ";")
  results_enrich_geneList$relative_enrich_random <- paste(results_enrich_randomGeneset$relative_enrich, sep = ";", collapse = ";")
  
  return(results_enrich_geneList)
}
```

## Enrichment analysis for each histone marks in each DE analysis

Compute enrichment for the 3 directions: all sex-DE genes, and separating male- and female-biased genes OR all DE genes and up- and down-regulated genes.

```{r init_enrichmentRes, echo=FALSE}
enrichRes <- data.frame("analysis" = NA, "direction" = NA, "nbr_DE" = NA, "nbr_geneList" = NA, 
                        "overlap" = NA, "N" = NA, "pval_enrich" = NA, "pval_deple" = NA, "relative_enrich" = NA,
                        "pval_enrich_random" = NA, "pval_deple_random" = NA, "relative_enrich_random" = NA,
                        "geneList")
enrichRes <- enrichRes[0,]
```

```{r enrich_, echo=FALSE}
directions <- c("both", "up", "down")
analyses <- list(sexDE_MG1_Xchr, sexDE_MG2_Xchr, sexDE_MG3_Xchr, DE_G1G2_Xchr, DE_G1G3_Xchr, DE_G3G2_Xchr)
analysesName <- c("MG1", "MG2", "MG3", "G1G2", "G1G3", "G3G2")
geneLists <- list(K27me3_normal_annot_Xchr_filtered, K27me3_eroded_annot_Xchr_filtered, K9me3_normal_annot_Xchr_filtered, K9me3_eroded_annot_Xchr_filtered)
geneListsName <- c("K27me3_normal", "K27me3_eroded", "K9me3_normal", "K9me3_eroded")
for (i in 1:length(analyses)){
  myAnalysis <- analyses[[i]]
  myAnalysisName <- analysesName[i]
  print(paste0(">>> ", myAnalysisName))
  for (j in 1:length(geneLists)){
    myGeneList <- geneLists[[j]]
    myGeneListName <- geneListsName[j]
    print(paste0(">> ", myGeneListName))
    for (myDirection in directions){
      print(myDirection)
      resTmp <- enrichment_analysis_permut(DEresults = myAnalysis, geneList = myGeneList, analysisName = myAnalysisName, 
                                           nbr_permut = 1000, pval_threshold = 0.05, direction = myDirection)
      # add name of the gene list analyzed
      resTmp$geneList <- myGeneListName
      # for sexDE analysis rename the direction 
      if (myAnalysisName %in% c("MG1", "MG2", "MG3")){
        if (myDirection == "up"){
          resTmp$direction <- "male"
        } else if (myDirection == "down"){
          resTmp$direction <- "female"
        }
      }
      # add results to the global enrichment abject
      enrichRes <- rbind(enrichRes, resTmp)
    }
  }
}
```

## Write results

Add percentage of the DE genes are binded by XIST.

```{r add_percent, echo=FALSE}
enrichRes$percent <- enrichRes$overlap/enrichRes$nbr_DE*100
```

Write the complete enrichment results into a file.

```{r write_res, echo=FALSE}
write.table(enrichRes, file = "enrichment_results.txt", quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t")
```


## Plot the enrichment results

Format a bit more the data

```{r format4plot_withPermut, echo=FALSE}
enrichRes_formated <- enrichRes
# create groups for facet
enrichRes_formated$analysisGroup <- ifelse(enrichRes_formated$analysis %in% c("MG1", "MG2", "MG3"), "sex-DE", "DE between female groups")
# add * for signif enrichment or depletion (with a threshold of 0.05)
enrichRes_formated$signif <- ifelse(enrichRes_formated$relative_enrich > 1, 
                                    ifelse(enrichRes_formated$pval_enrich<0.05, "*", ""),
                                    ifelse(enrichRes_formated$pval_deple<0.05, "*", ""))
# order the different direction of effect
enrichRes_formated$direction <- factor(enrichRes_formated$direction,
                                       levels = c("both", "female", "male", "down", "up"))

# extract and transform the results for the random set of genes
enrichRes_random_formated <- enrichRes_formated[, c("analysis", "direction", "analysisGroup", "geneList", "pval_enrich_random", "pval_deple_random", "relative_enrich_random")] %>% separate_longer_delim(cols = c(pval_enrich_random, pval_deple_random, relative_enrich_random), delim = ";")
enrichRes_random_formated$pval_enrich_random <- as.numeric(enrichRes_random_formated$pval_enrich_random)
enrichRes_random_formated$pval_deple_random <- as.numeric(enrichRes_random_formated$pval_deple_random)
enrichRes_random_formated$relative_enrich_random <- as.numeric(enrichRes_random_formated$relative_enrich_random)
```


```{r plot_enrich_withPermut, echo=FALSE}
p <- ggplot(enrichRes_random_formated[which(enrichRes_random_formated$direction != "both"), ], aes(x = analysis, y = relative_enrich_random, fill = direction, alpha = 0.4))
p <- p + geom_violin()
p <- p + geom_hline(yintercept = 1)
p <- p + scale_fill_manual(values = c("female" = "red", "male" = "blue", "down" = "black", "up" = "grey"))
p <- p + geom_point(data = enrichRes_formated[which(enrichRes_formated$direction != "both"), ], 
                    aes(x = analysis, y = relative_enrich, color = direction), position = position_dodge(0.9), size = 2, alpha = 1)
p <- p + geom_text(data = enrichRes_formated[which(enrichRes_formated$direction != "both"), ], 
                   aes(x = analysis, y = ifelse(relative_enrich>1, relative_enrich + 0.2, relative_enrich - 0.5), label = signif), position = position_dodge(0.9), size = 4, alpha = 1)
p <- p + scale_color_manual(values = c("female" = "red", "male" = "blue", "down" = "black", "up" = "grey"))
p <- p + facet_grid(cols = vars(analysisGroup), rows = vars(geneList), scales = "free")
p <- p + theme_minimal()
p <- p + xlab("") + ylab("Relative enrichment")
p <- p + guides(alpha="none")
p
# save plot
ggsave("enrichment_plot_permutation.pdf",
      units = "cm", width = 12, height = 12)
```

Plot only the results for the DE analysis between female groups.

```{r plot_enrich_withPermut_DEonly, echo=FALSE}
enrichRes_formated2 <- enrichRes_formated %>% separate_wider_delim("geneList", delim = "_", names = c("histoneMark", "cellStatus"))
enrichRes_formated2_select <- enrichRes_formated2[which(enrichRes_formated2$direction != "both" & enrichRes_formated2$analysisGroup != "sex-DE"), ]
enrichRes_formated2_select$cellStatus <- ifelse(enrichRes_formated2_select$cellStatus == "normal", "XaXi",
                                                ifelse(enrichRes_formated2_select$cellStatus == "eroded", "XaXe", ""))
enrichRes_random_formated2 <- enrichRes_random_formated %>% separate_wider_delim("geneList", delim = "_", names = c("histoneMark", "cellStatus"))
enrichRes_random_formated2_select <- enrichRes_random_formated2[which(enrichRes_random_formated2$direction != "both" & enrichRes_random_formated2$analysisGroup != "sex-DE"), ]
enrichRes_random_formated2_select$cellStatus <- ifelse(enrichRes_random_formated2_select$cellStatus == "normal", "XaXi",
                                                       ifelse(enrichRes_random_formated2_select$cellStatus == "eroded", "XaXe", ""))
p <- ggplot(enrichRes_random_formated2_select, aes(x = analysis, y = relative_enrich_random, fill = direction, alpha = 0.4))
p <- p + geom_violin()
p <- p + geom_hline(yintercept = 1)
p <- p + scale_fill_manual(values = c("female" = "red", "male" = "blue", "down" = "black", "up" = "grey"))
p <- p + geom_point(data = enrichRes_formated2_select, 
                    aes(x = analysis, y = relative_enrich, color = direction), position = position_dodge(0.9), size = 2, alpha = 1)
p <- p + geom_text(data = enrichRes_formated2_select, 
                   aes(x = analysis, y = ifelse(relative_enrich>1, relative_enrich + 0.2, relative_enrich - 0.5), label = signif), position = position_dodge(0.9), size = 4, alpha = 1)
p <- p + scale_color_manual(values = c("female" = "red", "male" = "blue", "down" = "black", "up" = "grey"))
#p <- p + facet_wrap(cellStatus~histoneMark, nrow = 2, ncol = 2, scales = "free")
p <- p + facet_grid(cols = vars(cellStatus), rows = vars(histoneMark))
p <- p + theme_minimal()
p <- p + xlab("") + ylab("Relative enrichment")
p <- p + coord_cartesian(ylim = c(0, 3)) 
p <- p + guides(alpha="none")
p
# save plot
ggsave("histoneMark_enrichment_plot_permutation_DEfemales.pdf",
      units = "cm", width = 12, height = 12)
```

