---
title: "Compare XIST binding genes to HipSci sex-DE genes"
author: "Clara Benoit-Pilven"
date: "`r Sys.Date()`"
output: html_document
---

```{r load_library, include=FALSE}
suppressPackageStartupMessages( library(readxl) )
suppressPackageStartupMessages( library(rtracklayer) )
suppressPackageStartupMessages( library(dplyr) )
suppressPackageStartupMessages( library(ggplot2) )
suppressPackageStartupMessages( library(tidyr) )
```

```{r set_workDir, echo=FALSE}
setwd("/Users/benoicla/Desktop/Workspace/iPSC/draft/revision/review_analyses/XISTbinding_analysis/")
```


# Load data

## Load XIST binding data

Load the list of genes overlapping with conserved autosomal XIST peaks in na√Øve human PSCs.

```{r load_XISTgenes, echo=FALSE}
XISTbinding_genes <- read_excel("Dror_2014_TableS4.xlsx", sheet = 1, skip = 4)
XISTbinding_genes$chr <- paste0("chr", XISTbinding_genes$chr)
XISTbinding_genes_prot <- XISTbinding_genes[which(XISTbinding_genes$`Gene type` %in% c("protein_coding", "lincRNA")), ]
```

## Load HipSci sex-DE genes

Load the 3 sets of sex-DE genes.

```{r load_hipsci_sexDE, echo=FALSE}
sexDE_MG1 <-  read_excel("Table_S7_toptable_ipsc_MG1.xlsx", sheet = 1)
sexDE_MG1_auto <- sexDE_MG1[which(!sexDE_MG1$chr %in% c("X", "Y", "MT")),]
sexDE_MG2 <-  read_excel("Table_S8_toptable_ipsc_MG2.xlsx", sheet = 1)
sexDE_MG2_auto <- sexDE_MG2[which(!sexDE_MG2$chr %in% c("X", "Y", "MT")),]
sexDE_MG3 <-  read_excel("Table_S9_toptable_ipsc_MG3.xlsx", sheet = 1)
sexDE_MG3_auto <- sexDE_MG3[which(!sexDE_MG3$chr %in% c("X", "Y", "MT")),]
```

Also load the different comparison betyween female groups.

```{r load_hipsci_femaleGrpDE, echo=FALSE}
DE_G1G3 <- read_excel("toptable_ipsc_G1G3.xlsx", sheet = 1)
DE_G1G3_auto <- DE_G1G3[which(!DE_G1G3$chr %in% c("X", "Y", "MT")),]
DE_G1G2 <- read_excel("toptable_ipsc_G1G2.xlsx", sheet = 1)
DE_G1G2_auto <- DE_G1G2[which(!DE_G1G2$chr %in% c("X", "Y", "MT")),]
DE_G3G2 <- read_excel("toptable_ipsc_G3G2.xlsx", sheet = 1)
DE_G3G2_auto <- DE_G3G2[which(!DE_G3G2$chr %in% c("X", "Y", "MT")),]
```


## Load Gencode 26 annotation

```{r load_annotation, echo=FALSE}
gencode26_annot <- rtracklayer::import("/Users/benoicla/Desktop/Genome/Gencode_v26/gencode.v26.annotation.gtf.gz")
gencode26 <- gencode26_annot[which(elementMetadata(gencode26_annot)$type == "gene"),]
gencode26 <- data.frame(gene_id = elementMetadata(gencode26)$gene_id,
                    gene_name = elementMetadata(gencode26)$gene_name,
                    gene_type = elementMetadata(gencode26)$gene_type,
                    chr = seqnames(gencode26),
                    start = start(gencode26),
                    end = end(gencode26),
                    strand = strand(gencode26))
```

# Enrichment analysis of XIST peaks in autosomal genes

## Checks and annotation

### Check

Check how many of the XIST binding genes are not found in the sex-DE results.

```{r check, echo=FALSE}
length(unique(XISTbinding_genes_prot$Gene))
length(intersect(XISTbinding_genes_prot$Gene, sexDE_MG1$gene_name))
```

The overlap is not as large as I was expecting: 313/573 genes.

### Format XIST peaks with our own annotation

For each peak in the XISTbinding_genes data frame, check if it overlap any gene in the sex-DE analysis MG1.

```{r annotate_XISTdata, echo=FALSE}
XISTbinding_genes_annotated <- XISTbinding_genes_prot
XISTbinding_genes_annotated$annotatedGene <- NA
for (i in 1:length(XISTbinding_genes_annotated$Gene)){
  chrPeak <- XISTbinding_genes_annotated$chr[i]
  startPeak <- XISTbinding_genes_annotated$`Peak start`[i]
  endPeak <- XISTbinding_genes_annotated$`Peak end`[i]
  # check if this peak overlap a gene from gencode 26 annotation by at least 1 base
  overlappingBool <- ifelse(chrPeak == gencode26$chr, 
                            ((startPeak<=gencode26$start & endPeak>=gencode26$start) | (endPeak>=gencode26$end & startPeak<=gencode26$end) | (startPeak>=gencode26$start & endPeak<=gencode26$end)),
                            FALSE)
  overlappingGenes <- gencode26[which(overlappingBool), ]
  # if only 1 gene, assigned this gene
  if (dim(overlappingGenes)[1] == 1){
    selectedGene <- overlappingGenes$gene_id
  } else if (dim(overlappingGenes)[1] > 1){
    # try to find the same gene that was annotated in the paper
    if (any(overlappingGenes$gene_name == XISTbinding_genes_annotated$Gene[i])){
      selectedGene <- overlappingGenes$gene_id[which(overlappingGenes$gene_name == XISTbinding_genes_annotated$Gene[i])]
    # select the protein coding gene if any
    } else if (any(overlappingGenes$gene_type == "protein_coding")){
      selectedGenes <- overlappingGenes$gene_id[which(overlappingGenes$gene_type == "protein_coding")]
      selectedGene <- selectedGenes[1]
    # otherwise take the first match
    } else{
      selectedGene <- overlappingGenes$gene_id[1]
    }
  } else if (dim(overlappingGenes)[1] == 0){
    selectedGene <- NA
  }
  XISTbinding_genes_annotated$annotatedGene[i] <- selectedGene
}
```

Check overlap now.
```{r check_new, echo=FALSE}
length(unique(XISTbinding_genes_annotated$annotatedGene))
length(intersect(XISTbinding_genes_annotated$annotatedGene, sexDE_MG1$gene_id))
```

The overlap is still quite small pointing to the fact that quite a number of XIST target are not analyzed in our dataset (e.g. not expressed in our data?!).

For the following analysis, only keep the XIST binding genes that were analyzed in our DE analysis.

```{r select_XISTbinding, echo=FALSE}
XISTbinding_genes_selected <- XISTbinding_genes_annotated[which(XISTbinding_genes_annotated$annotatedGene %in% sexDE_MG1$gene_id),]
XISTbinding_genes_selected_uniq <- unique(XISTbinding_genes_selected$annotatedGene)
```


## Enrichment of XIST binding genes in sex-DE genes?

```{r init_enrichmentRes, echo=FALSE}
enrichRes <- data.frame("analysis" = NA, "direction" = NA, "nbr_DE" = NA, "nbr_geneList" = NA, 
                        "overlap" = NA, "N" = NA, "pval_enrich" = NA, "pval_deple" = NA, "relative_enrich" = NA,
                        "pval_enrich_random" = NA, "pval_deple_random" = NA, "relative_enrich_random" = NA)
enrichRes <- enrichRes[0,]
```

### Functions

Create a function that take into input: 
- the list genes with the DE results, 
- the list of genes for which we want to know the enrichment,
- the p-value significant threshold [default = 0.05],
- the direction of effect (in the case we want to look at up and down-regulated genes separately) [default = both],

```{r function_enrichment_analysis, echo=FALSE}
enrichment_analysis <- function(DEresults, geneList, analysisName, pval_threshold = 0.05, direction = "both"){
  if (direction == "both"){
    DEresults_signif <- DEresults[which(DEresults$adj.P.Val < pval_threshold), ]
  } else if (direction == "up"){
    DEresults_signif <- DEresults[which(DEresults$adj.P.Val < pval_threshold & DEresults$logFC > 0), ]
  } else if (direction == "down"){
    DEresults_signif <- DEresults[which(DEresults$adj.P.Val < pval_threshold & DEresults$logFC < 0), ]
  }
  universe_nbr <- length(DEresults$gene_id)
  DE_nbr <- length(DEresults_signif$gene_id)
  geneList_nbr <- length(geneList)
  # overlap
  overlap_nbr <- length(intersect(DEresults_signif$gene_id, geneList))
  # get enrichment stat and effect size
  enrichment_stat_res <- enrichment_stat(overlap_nbr, geneList_nbr, DE_nbr, universe_nbr)
  # format data
  results <- data.frame("analysis" = analysisName, "direction" = direction, 
                        "nbr_DE" = DE_nbr, "nbr_geneList" = geneList_nbr, 
                        "overlap" = overlap_nbr, "N" = universe_nbr, 
                        "pval_enrich" = enrichment_stat_res$pvalEnrich, "pval_deple" = enrichment_stat_res$pvalDeple,
                        "relative_enrich" = enrichment_stat_res$relatEnrich)
  return(results)
}

enrichment_stat <- function(overlap_nbr, geneList_nbr, DE_nbr, universe_nbr){
  # overall enrichment
  #print(">> Enrichment")
  pvalEnrich <- phyper(overlap_nbr-1, geneList_nbr, universe_nbr - geneList_nbr, DE_nbr, lower.tail = FALSE)
  # overall depletion
  #print(">> Depletion")
  pvalDeple <-phyper(overlap_nbr,  geneList_nbr, universe_nbr -  geneList_nbr, DE_nbr, lower.tail = TRUE)
  # effect size
  #print(">> Relative enrichment")
  relatEnrich <- (overlap_nbr/ geneList_nbr)/(DE_nbr/universe_nbr)
  
  return(list("pvalEnrich" = pvalEnrich, "pvalDeple" = pvalDeple, "relatEnrich" = relatEnrich))
}
```

Function that do the enrichment analysis for the geneList of interest and a choosen number of random genesets.

```{r function_enrichment_analysis_permut, echo=FALSE}
enrichment_analysis_permut <- function(DEresults, geneList, analysisName, nbr_permut = 1000, pval_threshold = 0.05, direction = "both"){
  # enrichment for geneList
  results_enrich_geneList <- enrichment_analysis(DEresults = DEresults, geneList = geneList, 
                                                 analysisName = analysisName, pval_threshold = pval_threshold,
                                                 direction = direction)
  # enrichment for 'nbr_permut" random genesets of the same size of the geneList
  length_geneList <- length(geneList)
  results_enrich_randomGeneset <- results_enrich_geneList[0,]
  for (i in 1:nbr_permut){
    # select 'length_geneList' random genes from the universe
    random_geneset <- sample(DEresults$gene_id, length_geneList)
    # enrichment for this random geneset
    results_enrich_randomGeneset_tmp <- enrichment_analysis(DEresults = DEresults, geneList = random_geneset, 
                                                            analysisName = paste(analysisName, "random", sep = "_"), 
                                                            pval_threshold = pval_threshold, direction = direction)
    results_enrich_randomGeneset <- rbind(results_enrich_randomGeneset, results_enrich_randomGeneset_tmp)
  }
  results_enrich_geneList$pval_enrich_random <- paste(results_enrich_randomGeneset$pval_enrich, sep = ";", collapse = ";")
  results_enrich_geneList$pval_deple_random <- paste(results_enrich_randomGeneset$pval_deple, sep = ";", collapse = ";")
  results_enrich_geneList$relative_enrich_random <- paste(results_enrich_randomGeneset$relative_enrich, sep = ";", collapse = ";")
  
  return(results_enrich_geneList)
}
```



### MG1 analysis

Compute enrichment for the 3 directions: all sex-DE genes, and separating male- and female-biased genes.

```{r enrich_MG1, echo=FALSE}
directions <- c("both", "up", "down")
for (myDirection in directions){
  print(paste0(">>> ", myDirection))
  resTmp <- enrichment_analysis_permut(DEresults = sexDE_MG1_auto, geneList = XISTbinding_genes_selected_uniq, analysisName = "MG1", 
                                       nbr_permut = 1000, pval_threshold = 0.05, direction = myDirection)
  if (myDirection == "up"){
    resTmp$direction <- "male"
  } else if (myDirection == "down"){
    resTmp$direction <- "female"
  }
  # add results to the global enrichment abject
  enrichRes <- rbind(enrichRes, resTmp)
}
```


### MG2 analysis

Compute enrichment for the 3 directions: all sex-DE genes, and separating male- and female-biased genes.

```{r enrich_MG2, echo=FALSE}
directions <- c("both", "up", "down")
for (myDirection in directions){
  print(paste0(">>> ", myDirection))
  resTmp <- enrichment_analysis_permut(sexDE_MG2_auto, XISTbinding_genes_selected_uniq, "MG2", 
                                       nbr_permut = 1000, pval_threshold = 0.05, direction = myDirection)
  if (myDirection == "up"){
    resTmp$direction <- "male"
  } else if (myDirection == "down"){
    resTmp$direction <- "female"
  }
  # add results to the global enrichment abject
  enrichRes <- rbind(enrichRes, resTmp)
}
```


### MG3 analysis

Compute enrichment for the 3 directions: all sex-DE genes, and separating male- and female-biased genes.

```{r enrich_MG3, echo=FALSE}
directions <- c("both", "up", "down")
for (myDirection in directions){
  print(paste0(">>> ", myDirection))
  resTmp <- enrichment_analysis_permut(sexDE_MG3_auto, XISTbinding_genes_selected_uniq, "MG3", 
                                       nbr_permut = 1000, pval_threshold = 0.05, direction = myDirection)
  if (myDirection == "up"){
    resTmp$direction <- "male"
  } else if (myDirection == "down"){
    resTmp$direction <- "female"
  }
  # add results to the global enrichment abject
  enrichRes <- rbind(enrichRes, resTmp)
}
```


### G1G3 analysis

Compute enrichment for the 3 directions: all DE genes, and separating up- and down-biased genes.

```{r enrich_G1G3, echo=FALSE}
directions <- c("both", "up", "down")
for (myDirection in directions){
  print(paste0(">>> ", myDirection))
  resTmp <- enrichment_analysis_permut(DE_G1G3_auto, XISTbinding_genes_selected_uniq, "G1G3", 
                                       nbr_permut = 1000, pval_threshold = 0.05, direction = myDirection)
  # add results to the global enrichment abject
  enrichRes <- rbind(enrichRes, resTmp)
}
```


### G1G2 analysis

Compute enrichment for the 3 directions: all DE genes, and separating up- and down-biased genes.

```{r enrich_G1G2, echo=FALSE}
directions <- c("both", "up", "down")
for (myDirection in directions){
  print(paste0(">>> ", myDirection))
  resTmp <- enrichment_analysis_permut(DE_G1G2_auto, XISTbinding_genes_selected_uniq, "G1G2", 
                                       nbr_permut = 1000, pval_threshold = 0.05, direction = myDirection)
  # add results to the global enrichment abject
  enrichRes <- rbind(enrichRes, resTmp)
}
```



### G3G2 analysis

Compute enrichment for the 3 directions: all DE genes, and separating up- and down-biased genes.

```{r enrich_G3G2, echo=FALSE}
directions <- c("both", "up", "down")
for (myDirection in directions){
  print(paste0(">>> ", myDirection))
  resTmp <- enrichment_analysis_permut(DE_G3G2_auto, XISTbinding_genes_selected_uniq, "G3G2", 
                                       nbr_permut = 1000, pval_threshold = 0.05, direction = myDirection)
  # add results to the global enrichment abject
  enrichRes <- rbind(enrichRes, resTmp)
}
```


## Write results

Add percentage of the DE genes binded by XIST.

```{r add_percent, echo=FALSE}
enrichRes$percent <- enrichRes$overlap/enrichRes$nbr_DE*100
```


Write the complete enrichment results into a file.

```{r write_res, echo=FALSE}
write.table(enrichRes, file = "enrichment_results.txt", quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t")
```


## Plot the enrichment results

Plot the enrichment of XIST peaks with the enrichment seen in random set of genes

First, format a bit the data

```{r format4plot_withPermut, echo=FALSE}
enrichRes_formated <- enrichRes
# create groups for facet
enrichRes_formated$analysisGroup <- ifelse(enrichRes_formated$analysis %in% c("MG1", "MG2", "MG3"), "sex-DE", "DE between female groups")
# add * for signif enrichment or depletion (with a threshold of 0.05)
enrichRes_formated$signif <- ifelse(enrichRes_formated$relative_enrich > 1, 
                                    ifelse(enrichRes_formated$pval_enrich<0.05, "*", ""),
                                    ifelse(enrichRes_formated$pval_deple<0.05, "*", ""))
# order the different direction of effect
enrichRes_formated$direction <- factor(enrichRes_formated$direction,
                                       levels = c("both", "female", "male", "down", "up"))

# extract and transform the results for the random set of genes
enrichRes_random_formated <- enrichRes_formated[, c("analysis", "direction", "analysisGroup", "pval_enrich_random", "pval_deple_random", "relative_enrich_random")] %>% separate_longer_delim(cols = c(pval_enrich_random, pval_deple_random, relative_enrich_random), delim = ";")
enrichRes_random_formated$pval_enrich_random <- as.numeric(enrichRes_random_formated$pval_enrich_random)
enrichRes_random_formated$pval_deple_random <- as.numeric(enrichRes_random_formated$pval_deple_random)
enrichRes_random_formated$relative_enrich_random <- as.numeric(enrichRes_random_formated$relative_enrich_random)
```

Then, plot.

```{r plot_enrich2_withPermut, echo=FALSE}
p <- ggplot(enrichRes_random_formated[which(enrichRes_random_formated$direction != "both"), ], aes(x = analysis, y = relative_enrich_random, fill = direction, alpha = 0.4))
p <- p + geom_violin()
p <- p + geom_hline(yintercept = 1)
p <- p + scale_fill_manual(values = c("female" = "red", "male" = "blue", "down" = "black", "up" = "grey"))
p <- p + geom_point(data = enrichRes_formated[which(enrichRes_formated$direction != "both"), ], 
                    aes(x = analysis, y = relative_enrich, color = direction), position = position_dodge(0.9), size = 2, alpha = 1)
p <- p + geom_text(data = enrichRes_formated[which(enrichRes_formated$direction != "both"), ], 
                   aes(x = analysis, y = ifelse(relative_enrich>1, relative_enrich + 0.2, relative_enrich - 0.5), label = signif), position = position_dodge(0.9), size = 4, alpha = 1)
p <- p + scale_color_manual(values = c("female" = "red", "male" = "blue", "down" = "black", "up" = "grey"))
p <- p + facet_grid(cols = vars(analysisGroup), scales = "free", space = "free")
p <- p + theme_minimal()
p <- p + xlab("") + ylab("Relati ve enrichment")
p <- p + guides(alpha="none")
p
# save plot
ggsave("enrichment_plot_permutation2.pdf",
      units = "cm", width = 12, height = 6)
```

```{r plot_enrich2_withPermut_sexDE, echo=FALSE}
p <- ggplot(enrichRes_random_formated[which(enrichRes_random_formated$direction != "both" & enrichRes_random_formated$analysisGroup == "sex-DE"), ], aes(x = analysis, y = relative_enrich_random, fill = direction, alpha = 0.4))
p <- p + geom_violin()
p <- p + geom_hline(yintercept = 1)
p <- p + scale_fill_manual(values = c("female" = "red", "male" = "blue", "down" = "black", "up" = "grey"))
p <- p + geom_point(data = enrichRes_formated[which(enrichRes_formated$direction != "both" & enrichRes_formated$analysisGroup == "sex-DE"), ], 
                    aes(x = analysis, y = relative_enrich, fill = direction), position = position_dodge(0.9), size = 2, alpha = 1, pch=21, color = "black")
p <- p + geom_text(data = enrichRes_formated[which(enrichRes_formated$direction != "both" & enrichRes_formated$analysisGroup == "sex-DE"), ], 
                   aes(x = analysis, y = ifelse(relative_enrich>1, relative_enrich + 0.2, relative_enrich - 0.5), label = signif), position = position_dodge(0.9), size = 4, alpha = 1)
p <- p + scale_color_manual(values = c("female" = "red", "male" = "blue", "down" = "black", "up" = "grey"))
p <- p + theme_minimal()
p <- p + xlab("") + ylab("Relative enrichment")
p <- p + guides(alpha="none")
p
# save plot
ggsave("enrichment_plot_permutation2_sexDE.pdf",
      units = "cm", width = 8, height = 6)
```


```{r plot_enrich2_withPermut_femaleDE, echo=FALSE}
p <- ggplot(enrichRes_random_formated[which(enrichRes_random_formated$direction != "both" & enrichRes_random_formated$analysisGroup == "DE between female groups"), ], aes(x = analysis, y = relative_enrich_random, fill = direction, alpha = 0.4))
p <- p + geom_violin()
p <- p + geom_hline(yintercept = 1)
p <- p + scale_fill_manual(values = c("female" = "red", "male" = "blue", "down" = "black", "up" = "grey"))
p <- p + geom_point(data = enrichRes_formated[which(enrichRes_formated$direction != "both" & enrichRes_formated$analysisGroup == "DE between female groups"), ], 
                    aes(x = analysis, y = relative_enrich, fill = direction), position = position_dodge(0.9), size = 2, alpha = 1, pch=21, color = "black")
p <- p + geom_text(data = enrichRes_formated[which(enrichRes_formated$direction != "both" & enrichRes_formated$analysisGroup == "DE between female groups"), ], 
                   aes(x = analysis, y = ifelse(relative_enrich>1, relative_enrich + 0.2, relative_enrich - 0.5), label = signif), position = position_dodge(0.9), size = 4, alpha = 1)
p <- p + scale_color_manual(values = c("female" = "red", "male" = "blue", "down" = "black", "up" = "grey"))
p <- p + theme_minimal()
p <- p + xlab("") + ylab("Relative enrichment")
p <- p + guides(alpha="none")
p
# save plot
ggsave("enrichment_plot_permutation2_femaleDE.pdf",
      units = "cm", width = 8, height = 6)
```
